### PR TITLE
Fixes #33019 - Don't raise errors when deleting missing Candlepin environment

### PR DESCRIPTION
**Pull request by jturel**
There are two redmines associated to this issue in order to (hopefully) make cherry-picking back to 6.9 a little less annoying :)

(1) https://github.com/Katello/katello/pull/9270 introduced changes that will cause the Candlepin environment to be recreated on publish or promote if it's missing.

(2) This PR takes that slightly further and prevents a 404 from being raised when a Lifecycle Environment is deleted and for some reason the Candlepin environment is already gone.

A separate downstream PR will be needed to bring (1) back into a z-stream due to lots of refactored code.

**To test**

- Create a lifecycle environment
- Create a new CV, publish and promote while tailing candlepin.log
- See that a new environment was created:
``` 
2021-07-19 14:34:54,898 [thread=http-bio-127.0.0.1-23443-exec-5] [req=0ea37549-8fd7-45f8-836d-9f134a2ecab8, org=, csid=ed493181-9bd2-4805-b916-7d78113075f8] INFO  org.candlepin.servlet.filter.logging.LoggingFilter - Request: verb=POST, uri=/candlepin/owners/Default_Organization/environments
...
2021-07-19 14:34:55,890 [thread=http-bio-127.0.0.1-23443-exec-6] [req=ac55b2c4-d62d-4318-8e04-0a7c8f7e8779, org=, csid=ed493181-9bd2-4805-b916-7d78113075f8] INFO  org.candlepin.servlet.filter.logging.LoggingFilter - Request: verb=GET, uri=/candlepin/environments/e4f4f28fc7d767c47976a34e16c0c107
```

- Delete that environment from the rails console:
` Katello::Resources::Candlepin::Environment.destroy('e4f4f28fc7d767c47976a34e16c0c107')`
- Publish the CV again, and see that the environment is recreated by watching candlepin.log

- Delete the environment again from the Rails console (it should correspond to the LCE you created, not Library)
- Delete the Lifecyle Environment that you created and see that no errors get raised

### CODE REVIEW

File: app/lib/katello/errors.rb [hunk 1]
```diff
@@ -106,6 +106,7 @@ def self.from_exception(exception)
 
     class CandlepinNotRunning < StandardError; end
     class CandlepinPoolGone < CandlepinError; end
+    class CandlepinEnvironmentGone < CandlepinError; end
```
Comment (lfu) [inline app/lib/katello/errors.rb:109]: Could we have a generic class like `CandlepinResourceGone` that can be used for both pool and environment and all other kinds of resources in Candlepin? So we don't need to update this file again for similar issues from other resources.

**Comment by lfu** [comment]
> * Delete that environment from the rails console:
>   ` Katello::Resources::Candlepin::Environment.destroy('e4f4f28fc7d767c47976a34e16c0c107')`
> * Publish the CV again, and see that the environment is recreated by watching candlepin.log

The environment was recreated during `Promote` not `Publish`.

> * Delete the environment that you created and see that no errors get raised

**Before**
```
[2] pry(main)> Katello::Resources::Candlepin::Environment.destroy '212635834a99ec44cc20e94427971e7b'
2021-07-21T15:45:25 [D|kat|] Resource DELETE request: /candlepin/environments/212635834a99ec44cc20e94427971e7b
2021-07-21T15:45:25 [D|kat|] Headers: {"cp-user":"foreman_admin"}
2021-07-21T15:45:25 [D|kat|] Body: {}
/home/vagrant/katello/app/lib/katello/http_resource.rb:117: warning: URI.escape is obsolete
RestClient::NotFound: Katello::Resources::Candlepin::Environment: 404 Not Found {"displayMessage":"Environment with id 212635834a99ec44cc20e94427971e7b could not be found.","requestUuid":"31327170-19e6-4b25-ae62-d42db4e541ee"} (DELETE /candlepin/environments/212635834a99ec44cc20e94427971e7b)
from /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/rest-client-2.1.0/lib/restclient/abstract_response.rb:249:in `exception_with_response'
```
**After**
```
[2] pry(main)> Katello::Resources::Candlepin::Environment.destroy '212635834a99ec44cc20e94427971e7b'
2021-07-21T15:42:26 [D|kat|] Resource DELETE request: /candlepin/environments/212635834a99ec44cc20e94427971e7b
2021-07-21T15:42:26 [D|kat|] Headers: {"cp-user":"foreman_admin"}
2021-07-21T15:42:26 [D|kat|] Body: {}
/home/vagrant/katello/app/lib/katello/http_resource.rb:117: warning: URI.escape is obsolete
Katello::Errors::CandlepinEnvironmentGone: Katello::Resources::Candlepin::Environment: 404 Not Found {"displayMessage":"Environment with id 212635834a99ec44cc20e94427971e7b could not be found.","requestUuid":"2dae1bae-c129-4f09-8f5a-52fc866a49e1"} (DELETE /candlepin/environments/212635834a99ec44cc20e94427971e7b)
from /home/vagrant/katello/app/lib/katello/resources/candlepin/environment.rb:24:in `rescue in destroy'
Caused by RestClient::NotFound: Katello::Resources::Candlepin::Environment: 404 Not Found {"displayMessage":"Environment with id 212635834a99ec44cc20e94427971e7b could not be found.","requestUuid":"2dae1bae-c129-4f09-8f5a-52fc866a49e1"} (DELETE /candlepin/environments/212635834a99ec44cc20e94427971e7b)
from /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/rest-client-2.1.0/lib/restclient/abstract_response.rb:249:in `exception_with_response'
```
Is this as expected?

**Comment by jturel** [comment]
> > * Delete that environment from the rails console:
> >   ` Katello::Resources::Candlepin::Environment.destroy('e4f4f28fc7d767c47976a34e16c0c107')`
> > * Publish the CV again, and see that the environment is recreated by watching candlepin.log
> 
> The environment was recreated during `Promote` not `Publish`.
> 
> > * Delete the environment that you created and see that no errors get raised
> 
> **Before**
> 
> ```
> [2] pry(main)> Katello::Resources::Candlepin::Environment.destroy '212635834a99ec44cc20e94427971e7b'
> 2021-07-21T15:45:25 [D|kat|] Resource DELETE request: /candlepin/environments/212635834a99ec44cc20e94427971e7b
> 2021-07-21T15:45:25 [D|kat|] Headers: {"cp-user":"foreman_admin"}
> 2021-07-21T15:45:25 [D|kat|] Body: {}
> /home/vagrant/katello/app/lib/katello/http_resource.rb:117: warning: URI.escape is obsolete
> RestClient::NotFound: Katello::Resources::Candlepin::Environment: 404 Not Found {"displayMessage":"Environment with id 212635834a99ec44cc20e94427971e7b could not be found.","requestUuid":"31327170-19e6-4b25-ae62-d42db4e541ee"} (DELETE /candlepin/environments/212635834a99ec44cc20e94427971e7b)
> from /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/rest-client-2.1.0/lib/restclient/abstract_response.rb:249:in `exception_with_response'
> ```
> 
> **After**
> 
> ```
> [2] pry(main)> Katello::Resources::Candlepin::Environment.destroy '212635834a99ec44cc20e94427971e7b'
> 2021-07-21T15:42:26 [D|kat|] Resource DELETE request: /candlepin/environments/212635834a99ec44cc20e94427971e7b
> 2021-07-21T15:42:26 [D|kat|] Headers: {"cp-user":"foreman_admin"}
> 2021-07-21T15:42:26 [D|kat|] Body: {}
> /home/vagrant/katello/app/lib/katello/http_resource.rb:117: warning: URI.escape is obsolete
> Katello::Errors::CandlepinEnvironmentGone: Katello::Resources::Candlepin::Environment: 404 Not Found {"displayMessage":"Environment with id 212635834a99ec44cc20e94427971e7b could not be found.","requestUuid":"2dae1bae-c129-4f09-8f5a-52fc866a49e1"} (DELETE /candlepin/environments/212635834a99ec44cc20e94427971e7b)
> from /home/vagrant/katello/app/lib/katello/resources/candlepin/environment.rb:24:in `rescue in destroy'
> Caused by RestClient::NotFound: Katello::Resources::Candlepin::Environment: 404 Not Found {"displayMessage":"Environment with id 212635834a99ec44cc20e94427971e7b could not be found.","requestUuid":"2dae1bae-c129-4f09-8f5a-52fc866a49e1"} (DELETE /candlepin/environments/212635834a99ec44cc20e94427971e7b)
> from /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/rest-client-2.1.0/lib/restclient/abstract_response.rb:249:in `exception_with_response'
> ```
> 
> Is this as expected?

That looks correct. What I meant to say was: delete the lifecycle environment itself (via UI for example) which will trigger the deletion in Candlepin. Sorry for the confusion

**Comment by lfu** [comment]
> * Delete the Lifecyle Environment that you created and see that no errors get raised

Verified.

**Comment by jturel** [comment]
@jlsherrill can you verify the approach here - been verified by Lucy but wanna make sure it matches your understanding since I think you wrote up the kanboard card

File: app/lib/katello/errors.rb (hunk 1 above)
Comment (jturel) [reply app/lib/katello/errors.rb:109]: Let me think about that!

Comment (jturel) [reply app/lib/katello/errors.rb:109]: We could easily do that, but I don't think there's a value in doing it right now. At the moment, in each case, we want to know specifically that the Environment is not there or the Pool is not there. No use case right now for catching such errors generically, but would be easy to make them extend a generic class as you suggest in the future as they are laid out here.

**Comment by jlsherrill** [comment]
Yes, this matches what i was thinking about when i wrote the card +1 from me from a code perspective

**Comment by lfu** [comment]
LGTM 👍🏻

**Comment by jturel** [comment]
> LGTM 👍🏻

Awesome! Thank you @lfu . Can you "Approve" the PR so that I can merge it?

**Comment by lfu** [comment]
Approve!

**Review by lfu — APPROVED** [review]

**Review by lfu — APPROVED** [review]

**Comment by lfu** [comment]
[test katello]

**Comment by lfu** [comment]
APJ

**Comment by jturel** [comment]
@jlsherrill can you approve this as well?

**Pull request merged**